### PR TITLE
preliminary implementation of conditional logit model with outside op…

### PR DESCRIPTION
**Pull request by TianyuDu**
preliminary implementation of conditional logit model with outside option.

### CODE REVIEW

File: torch_choice/data/choice_dataset.py [hunk 1]
```diff
@@ -236,6 +252,12 @@ def num_sessions(self) -> int:
             # return the _num_sessions provided in the constructor.
             return self._num_sessions
         else:
+            num_unique = len(torch.unique(self.session_index))
+            expected_num_sessions = int(self.session_index.max()) + 1
+            if num_unique != expected_num_sessions:
+                warnings.warn(f"The number of sessions is inferred from the number of unique sessions in the session_index tensor. The session_index tensor in the ChoiceDataset ranges from {int(self.session_index.min())} to {int(self.session_index.max())}. The ChoiceDataset assumes session_index to be 0-indexed and encoded using consecutive integers. There are {expected_num_sessions} sessions expected given max(session_index). However, there are {num_unique} unique values in the session_index . This could be caused by missing sessions in the dataset (i.e., some sessions are not in session_index at all). If this is not expected, please check the session_index tensor. For a saver behavior, please provide the number of sessions explicitly by using the num_sessions keyword while initializing the ChoiceDataset class.")
```
Comment (kanodiaayush) [inline torch_choice/data/choice_dataset.py:258]: s/saver/safer

even better,
s/a saver/safer

File: torch_choice/data/choice_dataset.py [hunk 2]
```diff
@@ -236,6 +252,12 @@ def num_sessions(self) -> int:
...
+            else:
+                warnings.warn(f"The number of sessions is inferred from the number of unique sessions in the session_index tensor. This might lead to unexpected behaviors if some sessions never appeared in the session_index tensor. For a saver behavior, please provide the number of sessions explicitly by using the num_sessions keyword while initializing the ChoiceDataset class.")
```
Comment (kanodiaayush) [inline torch_choice/data/choice_dataset.py:260]: s/a saver/safer

File: torch_choice/model/conditional_logit_model.py [hunk 3]
```diff
@@ -106,6 +107,11 @@ def __init__(self,
                 Alternatively, users can pass a dictionary with keys exactly the same as the `coef_variation_dict` dictionary,
                 and values from {'normal', 'uniform', 'zero'} to initialize coefficients of different types of variables differently.
                 By default, all coefficients are initialized following a standard normal distribution.
+            model_outside_option (Optional[bool]): whether to explicitly model the outside option (i.e., the consumer did not buy anything).
+                To enable modeling outside option, the outside option is indicated by `item_index[n] == -1` in the item-index-tensor.
+                In this case, the item-index-tensor can contain values in `{-1, 0, 1, ..., num_items-1}`.
+                Otherwise, if the outside option is not modelled, the item-index-tensor should only contain values in `{0, 1, ..., num_items-1}`.
+                By default, model_outside_option is set to False and the model does not model the outside option.
```
Comment (kanodiaayush) [inline torch_choice/model/conditional_logit_model.py:114]: Add something like
"The utility of the outside option is always set to 0"

File: torch_choice/model/nested_logit_model.py [hunk 4]
```diff
@@ -91,6 +92,12 @@ def __init__(self,
 
             {nest, item}_weight_initialization (Optional[Union[str, Dict[str, str]]]): methods to initialize the weights of
                 coefficients for {nest, item} level model. Please refer to the `weight_initialization` keyword in ConditionalLogitModel's documentation for more details.
+
+            model_outside_option (Optional[bool]): whether to explicitly model the outside option (i.e., the consumer did not buy anything).
+                To enable modeling outside option, the outside option is indicated by `item_index[n] == -1` in the item-index-tensor.
+                In this case, the item-index-tensor can contain values in `{-1, 0, 1, ..., num_items-1}`.
+                Otherwise, if the outside option is not modelled, the item-index-tensor should only contain values in `{0, 1, ..., num_items-1}`.
+                By default, model_outside_option is set to False and the model does not model the outside option.
```
Comment (kanodiaayush) [inline torch_choice/model/nested_logit_model.py:100]: same; add that utility of outside option is always set to 0

**Comment by TianyuDu** [comment]
All comments have been resolved.

**Pull request merged**